### ISSUE TITLE
Feature: ML predictor embedding (add_predictor API)

### ISSUE THREAD

**Issue by jkitchin** [jkitchin/discopt#1]
## Summary

Add an `add_predictor()` API that automatically converts trained ML models into discopt constraints, enabling optimization over learned predictions. Inspired by [MathOptAI.jl](https://github.com/lanl-ansi/MathOptAI.jl) from LANL.

## Motivation

A common workflow in ML+optimization is: train a surrogate model (neural network, decision tree, GP), then optimize decisions subject to the surrogate's predictions. Today this requires manually unrolling model weights into constraints. An `add_predictor()` convenience layer would make this automatic.

discopt is well-positioned for this — the solver already handles MINLP, big-M (GDP), and nonlinear activations (McCormick/alphaBB relaxations). The missing piece is the import/conversion layer.

## Proposed API

```python
import discopt.modeling as dm
from discopt.ml import add_predictor

m = dm.Model("nn_optimization")
x = m.continuous("features", shape=(10,), lb=-1, ub=1)

# Embed a trained PyTorch/sklearn model as constraints
y, formulation = add_predictor(m, x, trained_model, method="bigm")

m.minimize(cost @ y)
m.subject_to(y[0] >= 0.8)
result = m.solve()
```

## Layer / Predictor Support

Priority tiers based on practical impact:

### Tier 1 — Core (neural networks)
- **Affine**: `W @ x + b` — trivial, already expressible
- **ReLU (big-M)**: binary variable + 4 constraints per neuron, maps to existing GDP machinery
- **ReLU (SOS1)**: SOS1 formulation alternative
- **Sigmoid / Tanh / SoftPlus**: nonlinear equality constraints, already supported in expression DAG
- **Pipeline**: chain layers sequentially

### Tier 2 — Tree ensembles
- **BinaryDecisionTree**: binary variable per leaf, one-hot selection constraints (Misic 2020)
- **Gradient boosting (XGBoost/LightGBM)**: decompose into per-tree formulations + sum

### Tier 3 — Advanced
- **Conv2d / AvgPool2d / MaxPool2d**: spatial layers with index bookkeeping
- **LayerNorm / Scale**: normalization layers
- **GELU / LeakyReLU / SoftMax**: additional activations
- **GCNConv / TAGConv**: graph neural network layers
- **Gaussian Processes**: mean + variance constraints

### Tier 4 — Interop
- **PyTorch import**: extract weights from `torch.nn.Module`
- **sklearn import**: extract from `sklearn` estimators (linear models, trees, MLPRegressor)
- **ONNX import**: universal model exchange format

## What Already Exists in discopt

- Big-M GDP reformulation (`gdp_reformulate.py`) — can encode ReLU
- Nonlinear expression support — sigmoid, tanh, exp in the DAG
- McCormick + alphaBB relaxations — handles nonlinear activations in the solver
- ICNN module (`icnn.py`) — convex neural network architecture (related but different use case)
- GNN branching (`gnn_branching.py`) — GNN infrastructure exists, though for branching not embedding

## Potential Advantages Over MathOptAI.jl

- **Solver-aware formulations**: discopt could exploit ReLU structure for branching priority or bound tightening, since the modeling layer and solver are co-designed
- **JAX autodiff**: automatic gradients through embedded layers without finite differences
- **Batch evaluation**: vmap over neural network constraint evaluations during B&B

## References

- MathOptAI.jl: https://github.com/lanl-ansi/MathOptAI.jl
- Misic (2020), "Optimization of Tree Ensemble Models" — decision tree MILP encoding
- Anderson et al. (2020), "Strong Mixed-Integer Programming Formulations for Trained Neural Networks"
- Fischetti & Jo (2018), "Deep Neural Networks and Mixed Integer Linear Optimization"
